golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -0,0 +1,421 @@
+const ExportedFilesMode = 0644
+
+var artifactNameSanitizer = regexp.MustCompile(`[^-_.a-zA-Z0-9]`)
+
+type activeSnapshot struct {
+	height        int64
+	startedResult chan error
+	retrieved     bool
+}
+
+type exportManifest struct {
+	BlockHeight uint64      `json:"blockHeight,omitempty"`
+	Data        string      `json:"data,omitempty"`
+	Artifacts   [][2]string `json:"artifacts"`
+}
+
+type SwingStoreExporter interface {
+	ExportSwingStore(ctx sdk.Context) []*vstoragetypes.DataEntry
+}
+
+type SwingsetSnapshotter struct {
+	app            *baseapp.BaseApp
+	logger         log.Logger
+	exporter       SwingStoreExporter
+	blockingSend   func(action vm.Jsonable) (string, error)
+	activeSnapshot *activeSnapshot
+}
+
+type snapshotAction struct {
+	Type        string            `json:"type"` // COSMOS_SNAPSHOT
+	BlockHeight int64             `json:"blockHeight"`
+	Request     string            `json:"request"` // "initiate", "discard", "retrieve", or "restore"
+	Args        []json.RawMessage `json:"args,omitempty"`
+}
+
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, exporter SwingStoreExporter, sendToController func(bool, string) (string, error)) SwingsetSnapshotter {
+	// The sendToController performed by this submodule are non-deterministic.
+	// This submodule will send messages to JS at unpredictable times from a
+	// goroutine. However the JS side does not perform consensus operations when
+	// handling these messages, so it is safe.
+
+	blockingSend := func(action vm.Jsonable) (string, error) {
+		bz, err := json.Marshal(action)
+		if err != nil {
+			return "", err
+		}
+		return sendToController(true, string(bz))
+	}
+
+	return SwingsetSnapshotter{
+		app:            app,
+		logger:         app.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "snapshotter"),
+		exporter:       exporter,
+		blockingSend:   blockingSend,
+		activeSnapshot: nil,
+	}
+}
+
+// Synchronously initiates a snapshot for the given height. If a snapshot is
+// already in progress, or if no snapshot manager is configured, this will fail.
+// The snapshot operation is performed in a goroutine, and synchronized with the
+// main thread through the `WaitUntilSnapshotStarted` method.
+func (snapshotter *SwingsetSnapshotter) InitiateSnapshot(height int64) error {
+	if snapshotter.activeSnapshot != nil {
+		return fmt.Errorf("snapshot already in progress for height %d", snapshotter.activeSnapshot.height)
+	}
+
+	if snapshotter.app.SnapshotManager() == nil {
+		return fmt.Errorf("snapshot manager not configured")
+	}
+
+	// Set the structure used to synchronize with the goroutine
+	// its presence indicates a snapshot has started.
+	snapshotter.activeSnapshot = &activeSnapshot{
+		height,
+		make(chan error, 1),
+		false,
+	}
+
+	go func(startedResult chan error) {
+		action := &snapshotAction{
+			Type:        "COSMOS_SNAPSHOT",
+			BlockHeight: snapshotter.activeSnapshot.height,
+			Request:     "initiate",
+		}
```

Comment:
I suggest introducing a `height` variable for use in logger output below.
```suggestion
		height := snapshotter.activeSnapshot.height

		action := &snapshotAction{
			Type:        "COSMOS_SNAPSHOT",
			BlockHeight: height,
			Request:     "initiate",
		}
```

Reply:
decided to create a dedicated logger for the active snapshot instead